File: device_exercises/simple_device_access_acc_org.cpp
#include <string>

#include <iostream>
#include <iomanip>
#include <sstream>

#include <unistd.h>

#include <sys/ioctl.h>
#include <linux/i2c-dev.h>
#include <fcntl.h>
#include <math.h>

int main(int argc, char* argv[])
{
    int m_accelerometer;
    //m_accelerometer = ::open("/dev/i2c-1", O_RDWR);
    m_accelerometer = ::open("/dev/i2c-1", O_RDWR);
    if(m_accelerometer < 0)
      {
        std::cout<<"Faild to open i2c port"<<std::endl;
        return false;
      }
    if (ioctl(m_accelerometer, I2C_SLAVE, 0x1d) < 0)
      {
        std::cout<<"Unable to get bus access to talk to slave"<<std::endl;
        return false;
      }
    ::usleep(100000);

    char buf[256];

    //POWER_CTL
    //buf[0] = 0x2d;buf[1] = 0x08;
    buf[0] = 0x2d;
    buf[1] = 0x08;
    ::write(m_accelerometer, buf, 2);

    //DATA_FORMAT
    //buf[0] = 0x31;buf[1] = 0x0b;
    buf[0] = 0x31;
    buf[1] = 0x0b;
    ::write(m_accelerometer, buf, 2);

    buf[0] = 0x32;
    ::write(m_accelerometer, buf, 1);
    ::read(m_accelerometer, buf, 6);
    
    short x_val = (((short)buf[1]<<8)|((short)buf[0]));
    short y_val = (((short)buf[3]<<8)|((short)buf[2]));
    short z_val = (((short)buf[5]<<8)|((short)buf[4]));

    std::stringstream ss;
    ss << (double)x_val*0.0039 << "," << (double)y_val*0.0039 << "," << (double)z_val*0.0039;

    std::cout << ss.str() << std::endl;

    ::close(m_accelerometer);
}

File: device_exercises/simple_device_access_acc_org.py
#!/usr/bin/env python
# -*- coding: euc-jp -*- 


##
# @file simple_device_access_acc_org.py
# @brief 

import io
import fcntl
import time
import struct

def main():

  I2C_SLAVE =      0x0703

  rp = io.open("/dev/i2c-1", mode="rb", buffering=0);
  wp = io.open("/dev/i2c-1", mode="wb", buffering=0);

  fcntl.ioctl(rp, I2C_SLAVE, 0x1d)
  fcntl.ioctl(wp, I2C_SLAVE, 0x1d)

  time.sleep(0.1);

  #POWER_CTL
  buf = bytearray() 
  buf.append(0x2d)
  buf.append(0x08)
  wp.write(buf);

  #DATA_FORMAT
  buf = bytearray() 
  buf.append(0x31)
  buf.append(0x0b)
  wp.write(buf);

  wp.write(bytearray([0x32]));
  data = rp.read(6);

  x_val = struct.unpack('<h', bytes(data[0:2]))[0]
  y_val = struct.unpack('<h', bytes(data[2:4]))[0]
  z_val = struct.unpack('<h', bytes(data[4:6]))[0]

  print(str(float(x_val)*0.0039) + "," + str(float(y_val)*0.0039) + "," + str(float(z_val)*0.0039))


  rp.close()
  wp.close()


    



if __name__ == '__main__':
  main()
